client/allocrunner/taskrunner/stats_hook.go

```diff
@@ -88,56 +88,18 @@ func (h *statsHook) Exited(context.Context, *interfaces.TaskExitedRequest, *inte
 // Collection ends when the passed channel is closed
 func (h *statsHook) collectResourceUsageStats(ctx context.Context, handle interfaces.DriverStats) {
 
-	ch, err := handle.Stats(ctx, h.interval)
+MAIN:
+	ch, err := h.callStatsWithRetry(ctx, handle)
 	if err != nil {
-		// Check if the driver doesn't implement stats
-		if err.Error() == cstructs.DriverStatsNotImplemented.Error() {
-			h.logger.Debug("driver does not support stats")
-			return
-		}
-		h.logger.Error("failed to start stats collection for task", "error", err)
+		return
 	}
 
-	var backoff time.Duration
-	var retry int
-	limit := time.Second * 5
 	for {
-		time.Sleep(backoff)
 		select {
 		case ru, ok := <-ch:
-			// Channel is closed
+			// if channel closes, re-establish a new one
 			if !ok {
-				var re *structs.RecoverableError
-				ch, err = handle.Stats(ctx, h.interval)
-				if err == nil {
-					goto RETRY
-				}
-
-				// We do not log when the plugin is shutdown since this is
-				// likely because the driver plugin has unexpectedly exited,
-				// in which case sleeping and trying again or returning based
-				// on the stop channel is the correct behavior
-				if err != bstructs.ErrPluginShutdown {
-					h.logger.Debug("error fetching stats of task", "error", err)
-					goto RETRY
```

Comment:
In my rewrite, I flipped this conditional logic.  The comment doesn't match the conditional.  I assume the comment intends to say that if the plugin is shutdown, we want to log in debug level rather than an error level (done in L130).  If so, the check here should be `err == bstructs.ErrPluginShutdown`

Reply:
I think the comment is correct - its saying that if the err is `ErrPluginShutdown` we should not be logging because that logging is unnecessary. I am not sure about retrying here vs one layer up to call `TaskStats` once the plugin restarts though.

Not an expert in this code though, so @nickethier should chime in..

Reply:
In the old code, if `err != bstructs.ErrPluginShutdown` we log in debug mode in L121, but if it's `ErrPluginShutdown`, then we will log with a Warn level in L130.  I'm having a hard time reconciling that with comment.

The old logic for logging is in: https://github.com/hashicorp/nomad/blob/8da0b91287023d8a410fc101f39b821e1910e28e/client/allocrunner/taskrunner/stats_hook.go#L120-L130